test/affprop.jl

```diff
@@ -46,6 +46,14 @@ using Statistics
         @test R.counts[i] == count(==(i), R.assignments)
     end
 
+    @testset "Ensure works on nonbasic array type" begin
+        R2 = affinityprop(@view S[:,:])  # run on complete subarray
```

Comment:
It's as long as the previous version, but a bit more explicit (and without macro magic ;)).
Could you please update it in the other tests as well?
```suggestion
        R2 = affinityprop(view(S, 1:n, 1:n))  # run on complete subarray
```

Reply:
That is a strictly worse change.

`@view S[:, :]` is the same as `view(S, :, :)`. *not* the same as `@view(S, 1:n, 1:n)`
Putting in the `1:n` makes it hit different a different code path.
Now it has to create a `Base.OneTo(n)` object and do a bounds-check.
So it is slower.
```
julia> using BenchmarkTools

julia> const d = rand(1000,1000);

julia> const n = size(d,1);

julia> @btime(@view ($d)[:,:]);
  6.964 ns (1 allocation: 48 bytes)

julia> @btime(view($d, 1:$n, 1:$n));
  8.807 ns (1 allocation: 64 bytes)

julia> @btime(view($d, :, :));
  6.875 ns (1 allocation: 48 bytes)
```

Its also more confusing to read since I need to go and find what `n` was defined to be.
Where as `:` means what it always means; the whole dimension.

Further more using `1:n` breaks for OffsetArrays etc.
so its a bad pattern to follow.

Reply:
> That is a strictly worse change.

Obviously, I don't recommend using it in the actual code.
The intention is to create `AbstractMatrix` that is more different from `Matrix` than `view(M, :, :)`, so that -- if the clustering method does not fully support `AbstractArray` in some way -- we have a bit more chances to detect it.
From that perspective `view(M, row_ixs, col_ixs)` or `sparse()` would be even better, since it would test whether the clustering method relies on stride.

But I'm not bound to this suggestion, especially if `1:n` hits the performance of some tests hardly.
I guess we will be adding some type checks/warnings as the users will discover the limitations of `::AbstractMatrix`, so it's fine even with these very basic tests.

Reply:
> hits the performance of some tests hardly.

It's not about performance here, it is about being consistent with the code you would write else where. And everywhere you would use `:` since in other places performance *does* mater.

> IMHO using @view does make sense to convert multiple indexing operations at once or to turn views on/off, otherwise it looks confusing.

`@view` does not do that. You are thinking of `@views` which applies to a single block of code.
`@view` only applies to a single indexing expression, not to a block of code.
`@view` makes sure that writing `xs[5:end]` works, since can't do `@view

---

> From that perspective view(M, row_ixs, col_ixs) or sparse() would be even better, since it would test whether the clustering method relies on stride.

If that is your goal then we can't use `view(x, 1:n, 1:n)` either since `strides` works fine on that. we will need to densify the indexed.
Which can be written clearest as `@view x[collect(1:end), collect(1:end)]`
but since I can't use the macro so.

Reply:
> @view does not do that. You are thinking of @views which applies to a single block of code.

My bad, thanks for clarification.

I hope we understood each other.
If you want to go deeper with testing various `AbstractMatrix` corner cases, you are welcome to do so.
I was just pointing you to this direction with my original comment on `view(S, 1:n, 1:n)`, but the possibilities are endless, and I'm fine with `view(S, :, :)` as a starting point.

But, please, change to `view(S, ...)` instead of `@view` macro.
Yes, it could be regarded as a matter of personal preference, but it's also about the code consistency, since in the other parts of the package we use `view()`.

Thank you!

Reply:
this is done